### PR TITLE
Install all dependencies on Linux

### LINKED ISSUE #2548: Need an easy setup script for linux boxes
### User Story

As a developer, I want to be able to get going quickly with Status on my Linux machine so that I don't get blocked by trivial things I shouldn't need to worry about before getting my hands really dirty.

### Description

*Type*: Feature

*Summary*: I would love @dbalatero (or any other contributor) to write a bash script for Linux that devs can use to build Status easily.

#### Expected behavior
Easy as 1, 2, 3 to setup and build Status on a Linux box.

#### Actual behavior
Currently, you need to read the wiki and follow a lot of steps with plenty of room for error, which can be discouraging.

### Solution
*Summary*: 

- Farm it out to Open Bounty and let our amazing community get to work!

### Additional Information
[comment]: # (Please do your best to fill this out.)
* Status version: 0.9.12+
[comment]: # (Android or iOS.)
* Operating System: Linux

### PR DESCRIPTION
**Pull request by dbalatero**
Fixes #2548 

### Summary:

This merges in Linux-specific setup to the `scripts/setup` script, and renames it from `setup-osx` to `setup`.

### Review notes:

* This only is tested in Ubuntu 16.04 LTS Desktop
* I haven't implemented installing the Android SDK, as this is hairier than anticipated. I think that might be better done in a separate PR. That install in particular seems like the worst one to deal with, and I think iterating on this to get a solid Android setup would be good.

### Steps to test:

1. Clone the repo on ubuntu.
2. Run `scripts/setup`
3. It should work!

status: ready

### CODE REVIEW

**Comment by charlesg3** [comment]
ok great -- I'll take a look at coming up with a solution that runs the setup on several flavors of linux using docker.

Is there any precedent for using docker in this project? I can try to actually install docker as well as a part of the flow, but that may be out of scope (I sort of prefer the scheme of -- see if support is provided, if not install it... e.g. batteries included.)

Just let me know and I'll try to run with it.

**Comment by dbalatero** [comment]
@charlesg3 I don't think Docker is integrated on this project, so this would be a question for the Status core team.

**Comment by dbalatero** [comment]
@yenda Maybe you can help @charlesg3 with his question?

**Comment by dbalatero** [comment]
@flexsurfer @pablanopete Hey just want to wrap up a few outstanding PRs before I fly out on holiday vacation Friday morning. Let me know what I need to do on this!

**Comment by flexsurfer** [comment]
@jeluard or @goranjovic  could you please take a look

**Comment by jeluard** [comment]
@dbalatero Not sure I want to risk messing with my env.. ;) Is this script supposed to upgrade existing tools to latest expected? Maybe we could test with a small docker image?
Also what should I expect after running those: everything is available but `avd` ?

**Comment by dbalatero** [comment]
@jeluard yes, ideally it will not mess up your env, and just install things that are missing and/or upgrade package manifests via maven, npm, etc.

I think it should be safe to run. Just do a quick audit of the script to put yourself at ease.

Here is the main flow:

https://github.com/dbalatero/status-react/blob/cb32d8e5d329422a94de46759c9d8e737c34e687/scripts/setup#L24-L47

And here is where each of those installer functions are defined:

https://github.com/dbalatero/status-react/blob/cb32d8e5d329422a94de46759c9d8e737c34e687/scripts/lib/setup/installers.sh

**Comment by dbalatero** [comment]
> Also what should I expect after running those: everything is available but avd

Correct, everything but Android tools at this time.

File: scripts/setup [hunk 1]
```diff
@@ -4,38 +4,33 @@
 # This install script will setup your development dependencies on OS X.
 # Windows/Linux support is welcome!
```
Comment (charlesg3) [inline scripts/setup]: This comment should be updated since the PR adds support for linux.

**Comment by charlesg3** [comment]
Which version of ubuntu is this meant to work against? I am using Docker to test the setup in a clean environment.

**Comment by dbalatero** [comment]
@charlesg3 Updated the script comment to add Ubuntu.

> Which version of ubuntu is this meant to work against? I am using Docker to test the setup in a clean environment.

I built it on 16.04. Maybe verify it works on 16.04 as well, and then you can try 14.04? In general it should be portable but some packages with weird build processes might have different PPA repositories depending on version, etc…

**Comment by charlesg3** [comment]
Great -- I'm building against that version.

Is there a test suite I can run after doing setup to ensure that everything is actually setup? (`lein test` ?)

**Comment by dbalatero** [comment]
@charlesg3 That's a great question–I feel like a good test would be to build and run in a simulator, but this PR doesn't include bootstrapping the Android SDK, which would be a pre-req for building the app :\

Maybe this needs to get the SDK in as well to actually complete the PR and test it?

**Comment by dbalatero** [comment]
(installing the SDK is something I'm still baffled about, btw)

**Comment by charlesg3** [comment]
Ok makes sense... I think that for now perhaps just ensuring that the appropriate components are installed is a good stopping point for this particular PR. To keep things manageable, I'd like to see new / separate issues for:
- support for testing this with docker (quite simple), perhaps multiple flavors (17.10 as well?)
- support installing the SDK as well during setup (perhaps this is separate / optional)?
- testing the full suite with the SDK from the docker container
- breaking out the base docker container from the source code so that it's possible to pull down a container that has all of the deps pre-installed and then add the source code to that image (esentially a development container)

As long as this doesn't actually break anything existing (and adds support for more functionality). Then I do think it should be able to land as-is. I'll report when the full install finishes on ubuntu 16.04.

I'm pretty new to the project here so someone else will have to comment on what should actually happen / where / how to break things up.

**Comment by charlesg3** [comment]
Another curiosity -- why is there a `postinstall.sh` and why isn't that just a part of the setup script (or at the very least run for you _by_ the setup script)?

**Comment by dbalatero** [comment]
> Another curiosity -- why is there a postinstall.sh and why isn't that just a part of the setup script (or at the very least run for you by the setup script)?

`postinstall.sh` is ran by `npm install`, see here: https://github.com/dbalatero/status-react/blob/experiment/switch-to-yarn%232541/package.json#L7

Some crazy stuff is happening regarding JS package management. The short of it is:

* react-native has a bug where you can't access core Node APIs (like filesystem, etc)
* `fs-nodeify` patches your `node_modules` to pull those back in
* for some unholy reason, `postinstall.sh` is upgrading `react` and `react-native-tcp` after `fs-nodeify` runs, instead of getting those committed to the lock file

I went down this hole a bit already, with this issue I raised: https://github.com/status-im/status-react/issues/2541

I have an experimental branch where I'm trying to replace `npm` with `yarn` and get a consistent lock file that doesn't require this hoop jumping. Unfortunately, it doesn't build yet. I think I need to do another careful pass and hand-resolve any npm -> yarn migration issues.

My branch: https://github.com/dbalatero/status-react/tree/experiment/switch-to-yarn%232541

My updated `postinstall.sh` script: https://github.com/dbalatero/status-react/blob/experiment/switch-to-yarn%232541/postinstall.sh

**Comment by dbalatero** [comment]
Also with that weird postinstall.sh react upgrade, it doesn't upgrade all the packages to the same react version, so dependencies are out of sync (`react-dom` needs to be upgraded as well, for example)

**Comment by charlesg3** [comment]
Ok -- thanks for the info.

Results from building in a "clean" docker container (had to install git, sudo and a few things that we can assume are installed by the user -- I will check in a script that tests this separately so that we can ensure that we don't regress the install):

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 05:19 min
[INFO] Finished at: 2017-12-06T01:22:04+00:00
[INFO] Final Memory: 16M/170M
[INFO] ------------------------------------------------------------------------

  + done

Setup complete!
===============

There are a few manual steps you might want to do:

1. Optionally set up Genymotion if you don't want to use Android Virtual Device:

    https://www.genymotion.com

2. Setup Android Development Environment + Simulator:

    https://facebook.github.io/react-native/docs/android-setup.html

3. Add your SSH public key to Github if it isn't already in there.
```

**Comment by charlesg3** [comment]
@jeluard ⬆️

**Comment by jeluard** [comment]
Great! @charlesg3 Did you try to run `./re-natal use-figwheel`, `lein test-cljs` and `lein prod-build`?

**Comment by andytudhope** [comment]
@dbalatero @jeluard `lein test-cljs` and `lein prod-build` both pass well with no issues. However, something seems to be the matter with `./re-natal use-figwheel`:
```
ubuntu@ip-172-31-30-207:~/status-react$ ./re-natal use-figwheel
-bash: ./re-natal: No such file or directory
```

Do I need to run the `postinstall.sh` script, because I did not do that...

**Comment by charlesg3** [comment]
@jeluard -- I will try those and report back

**Comment by charlesg3** [comment]
@jeluard  -- running from the built docker container...

```
root@4689f3455356:/status# ./re-natal use-figwheel
Cleaning...
index.ios.js was regenerated
index.android.js was regenerated
Host in RCTWebSocketExecutor.m was updated
Dev server host for iOS: localhost
Dev server host for Android: 10.0.3.2
```

```
root@4689f3455356:/status# lein test-cljs
Retrieving lein-doo/lein-doo/0.1.7/lein-doo-0.1.7.pom from clojars
Retrieving doo/doo/0.1.7/doo-0.1.7.pom from clojars
Retrieving karma-reporter/karma-reporter/0.1.0/karma-reporter-0.1.0.pom from clojars
Retrieving org/clojure/clojurescript/0.0-3308/clojurescript-0.0-3308.pom from central
Retrieving org/clojure/clojure/1.7.0-RC1/clojure-1.7.0-RC1.pom from central
Retrieving com/google/javascript/closure-compiler/v20150505/closure-compiler-v20150505.pom from central
Retrieving com/google/javascript/closure-compiler-parent/v20150505/closure-compiler-parent-v20150505.pom from central
Retrieving com/google/javascript/closure-compiler-externs/v20150505/closure-compiler-externs-v20150505.pom from central
Retrieving org/clojure/google-closure-library/0.0-20150505-021ed5b3/google-closure-library-0.0-20150505-021ed5b3.pom from central
Retrieving org/clojure/google-closure-library-third-party/0.0-20150505-021ed5b3/google-closure-library-third-party-0.0-20150505-021ed5b3.pom from central
Retrieving org/clojure/tools.reader/0.9.2/tools.reader-0.9.2.pom from central
Retrieving doo/doo/0.1.7/doo-0.1.7.jar from clojars
Retrieving lein-doo/lein-doo/0.1.7/lein-doo-0.1.7.jar from clojars
Retrieving karma-reporter/karma-reporter/0.1.0/karma-reporter-0.1.0.jar from clojars
Retrieving day8/re-frame/test/0.1.5/test-0.1.5.pom from clojars
Retrieving org/clojure/clojurescript/1.9.89/clojurescript-1.9.89.pom from central
Retrieving re-frame/re-frame/0.9.3/re-frame-0.9.3.pom from clojars
Retrieving day8/re-frame/test/0.1.5/test-0.1.5.jar from clojars
WARNING: Protocol IFn implements method -invoke with variadic signature (&) at line 61 target/test/reagent/impl/util.cljs

;; ======================================================================
;; Testing with Node:

... [SNIP] ...

Ran 47 tests containing 299 assertions.
0 failures, 0 errors.
```

```
root@4689f3455356:/status# lein prod-build
Compiling ClojureScript...
Retrieving cljsbuild/cljsbuild/1.1.7/cljsbuild-1.1.7.pom from clojars
Retrieving org/clojure/clojurescript/0.0-3211/clojurescript-0.0-3211.pom from central
Retrieving org/clojure/clojure/1.7.0-beta1/clojure-1.7.0-beta1.pom from central
Retrieving clj-stacktrace/clj-stacktrace/0.2.5/clj-stacktrace-0.2.5.pom from clojars
Retrieving org/clojure/tools.namespace/0.2.11/tools.namespace-0.2.11.pom from central
Retrieving org/clojure/tools.namespace/0.2.11/tools.namespace-0.2.11.jar from central
Retrieving clj-stacktrace/clj-stacktrace/0.2.5/clj-stacktrace-0.2.5.jar from clojars
Retrieving cljsbuild/cljsbuild/1.1.7/cljsbuild-1.1.7.jar from clojars
Compiling ["index.ios.js"] from ["react-native/src" "src" "env/prod"]...
Successfully compiled ["index.ios.js"] in 86.268 seconds.
Compiling ClojureScript...
Compiling ["index.android.js"] from ["react-native/src" "src" "env/prod"]...
Successfully compiled ["index.android.js"] in 92.554 seconds.
```

**Comment by dbalatero** [comment]
That re-natal script is supposed to be getting symlinked from
postinstall.sh, hm.

**Comment by charlesg3** [comment]
@dbalatero -- the re-natal script _does_ seem to be available in the clean docker run (and runs successfully per my output above).

**Comment by dbalatero** [comment]
@andytudhope `postinstall.sh` is supposed to run after `npm install` by convention. If that's not happening, something weird is going on :\

**Comment by dbalatero** [comment]
@charlesg3 I think the `./re-natal` symlink should really just happen in `scripts/setup` and get moved out of the postinstall script. I'll update that if you agree :+1:

**Comment by charlesg3** [comment]
Yes I think that makes sense

File: scripts/lib/setup/installers.sh [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+      automake
+      build-essential
+      cmake
+      curl
+      libssl-dev
+      libtool
+      make
+      python-dev
+      wget
+      unzip
```
Comment (andytudhope) [inline scripts/lib/setup/installers.sh:29]: Add `g++` here. Seems to be a requirement on AWS Ubuntu 16.04 machines. For some reason `build-essential` doesn't seem to cover the bases required for `lzma-native` in particular.

Comment (dbalatero) [reply scripts/lib/setup/installers.sh:29]: Done

**Comment by dbalatero** [comment]
> Yes I think that makes sense

@charlesg3 Updated the symlinks.

**Comment by andytudhope** [comment]
@dbalatero could we check for error codes during the `npm install` particularly so that if it does fail on others' machines they're fully aware of it, rather than thinking the install has succeeded like the end of the script suggested to me the first time I ran it?

**Review by andytudhope — APPROVED** [review]

**Comment by jeluard** [comment]
@dbalatero BTW, any reason to use `apt-get` and not `apt` ?

**Comment by jeluard** [comment]
@dbalatero Actually we want to make sure Linux users not relying on `apt` are not left out. 
I would propose the following: 
* if you have all deps then all good
* if some deps are missing asks if it's ok to install with `apt`. If not print details about what must be installed and exit
* profit!

Sorry for not seeing this before.

**Comment by dbalatero** [comment]
@jeluard Is there an Ubuntu case where they aren't using `apt`? If no I'd suggest merging this, then we can iterate to add other flavors. I can switch `apt-get` to `apt` though.

**Comment by dbalatero** [comment]
@jeluard switched to `apt`

**Comment by dbalatero** [comment]
One stopgap we can do for now so we don't confuse non-Ubuntu non-apt users is:

```bash
function is_ubuntu() {
  !is_linux && return 1
  (lsb_release -a 2>/dev/null | grep -q Ubuntu) && (command -v apt >/dev/null)
}

if is_linux && !is_ubuntu; then
  cecho "@b@yellow[[Only Ubuntu Linux with apt is currently supported. \
Please install any packages manually: \
https://wiki.status.im/Building_Status]]"
  exit 1
fi
```

**Comment by jeluard** [comment]
@dbalatero Sounds good to me!

**Comment by dbalatero** [comment]
@jeluard Added a check to enforce Ubuntu only for now.

If nothing else is pending, let's get this merged! Otherwise let me know soon, I can make changes today before I leave town.

**Review by jeluard — APPROVED** [review]

**Review by yenda — APPROVED** [review]

**Comment by puilp0502** [comment]
@dbalatero Actually, apt is supposed to be used by humans, not scripts. [The man page](http://manpages.ubuntu.com/manpages/wily/en/man8/apt.8.html)(DIFFERENCES TO APT-GET(8) section)  states that `The apt command  ... does not need to be backward compatible like apt-get(8)`. So I think it will be better to stick with apt-get.

**Comment by jeluard** [comment]
@puilp0502 In our scripts output is not interpreted so this is not an issue.
`apt` provides better output, which we rely on here for humans, so it looks better to me.

**Comment by puilp0502** [comment]
@jeluard I was worried about the changes `apt` might bring that will break the script.

**Comment by charlesg3** [comment]
@puilp0502 --
My suspicion is that the warnings about apt not being backwards compatible is much more likely to be related to the output (as @jeluard alluded to) than the input interface (e.g. for us `apt install X`) or the package name (which needs to be consistent across both packages.

Furthermore, I think any changes to apt (as a system utility) will likely be infrequent and at distro boundaries, so any such changes are less likely to be "unexpected", and can be tested for using docker scripts that are in the pipeline.

Again, these are assumptions but I think _in practice_ it will be just as safe to use apt here as apt-get.

**Comment by puilp0502** [comment]
@charlesg3 Okay, This PR LGTM :)

**Comment by jeluard** [comment]
Thanks @dbalatero for the hard work!

**Pull request merged**